internal/admission/defaulting_test.go

```diff
@@ -111,3 +111,142 @@ func TestTimeout(t *testing.T) {
 		require.InDelta(t, timeout.Seconds(), time.Since(start).Seconds(), 0.1, "timeout duration is not respected")
 	})
 }
+
+func TestFlow(t *testing.T) {
+	//GIVEN
+	logger := zap.NewNop()
+	scheme := runtime.NewScheme()
+	require.NoError(t, corev1.AddToScheme(scheme))
+	decoder, err := admission.NewDecoder(scheme)
+	require.NoError(t, err)
+
+	testNs := "test-namespace"
+	ns := corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: testNs, Labels: map[string]string{
+		pkg.NamespaceValidationLabel: pkg.NamespaceValidationEnabled,
+	}}}
+
+	type args struct {
+		labels            map[string]string
+		shouldBeValidated bool
+	}
+	type want struct {
+		patchesCount     int
+		validationStatus string
+	}
+	tests := []struct {
+		name string
+		args args
+		want want
+	}{
+		{
+			name: "pod without label should pass with validation and set Success",
+			args: args{
+				labels:            nil,
+				shouldBeValidated: true,
+			},
+			want: want{
+				patchesCount:     1,
+				validationStatus: pkg.ValidationStatusSuccess,
+			},
+		},
+		{
+			name: "pod with label Success should pass with validation",
+			args: args{
+				labels:            map[string]string{pkg.PodValidationLabel: pkg.ValidationStatusSuccess},
+				shouldBeValidated: true,
+			},
+			want: want{
+				patchesCount: 0,
+			},
+		},
+		{
+			name: "pod with label Failed should pass without validation and unchanged",
+			args: args{
+				labels:            map[string]string{pkg.PodValidationLabel: pkg.ValidationStatusFailed},
+				shouldBeValidated: false,
+			},
+			want: want{
+				patchesCount: 0,
+			},
+		},
+		{
+			name: "pod with label Reject should pass without validation and unchanged",
+			args: args{
+				labels:            map[string]string{pkg.PodValidationLabel: pkg.ValidationStatusReject},
+				shouldBeValidated: false,
+			},
+			want: want{
+				patchesCount: 0,
+			},
+		},
+		{
+			name: "pod with label Pending should pass without validation and unchanged",
+			args: args{
+				labels:            map[string]string{pkg.PodValidationLabel: pkg.ValidationStatusPending},
+				shouldBeValidated: false,
+			},
+			want: want{
+				patchesCount: 0,
+			},
+		},
+		{
+			name: "pod with unknown label should pass without validation and unchanged",
+			args: args{
+				labels:            map[string]string{pkg.PodValidationLabel: "some-unknown-label"},
+				shouldBeValidated: false,
+			},
+			want: want{
+				patchesCount: 0,
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			//GIVEN
+			imageName := "test:should-not-be-validated"
+			if tt.args.shouldBeValidated {
+				imageName = "test:should-be-validated"
+			}
+			pod := corev1.Pod{ObjectMeta: metav1.ObjectMeta{Name: "pod", Namespace: testNs},
```

Comment:
Can we have test `arg` pod in testCases instead of creating it here?
We would get less ifs and easier to understand test case.

Reply:
Yes, but IMO when we pass image name or pod, test parameters reading will be not so clear